### PR TITLE
WithoutDatastore: coerce blockstore into datastore.

### PR DESCRIPTION
**Pull request by raulk**
This implements a constructor that makes the Filestore work entirely against a blockstore by using a coercer.

This allows us to leverage the CARv2 blockstore as a "self-contained virtual CAR", containing the inermediate blocks as-is, and the UnixFS leaves as positional mappings to the original file.

Without this coercion, we would still need a datastore (e.g. Badger) to store the positional mappings.

<sup>Yes, it's a hack, but this whole library/circuitry is.</sup>

### CODE REVIEW

File: blockstore.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	// PosInfo leaves to the datastore (which in our case is the coerced
+	// blockstore), and the intermediate nodes to the blockstore proper (since
+	// they cannot be mapped to the file.
+	fstore := NewFilestore(bs, fm)
+	bs = blockstore.NewIdStore(fstore)
+
+	return bs, nil
+}
+
+var cidBuilder = cid.V1Builder{Codec: cid.Raw, MhType: mh.IDENTITY}
```
Comment (raulk) [inline blockstore.go:35]: Should we use another multihash for more compactness?

**Review by Kubuxu — COMMENTED** [review]
The code looks correct. Whether we want to do it is another question.

File: blockstore.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// WithoutDatastore returns a FileManager/Filestore backed entirely by a
+// blockstore without requiring a datastore. It achieves this by coercing the
+// blockstore into a datastore. The resulting blockstore is suitable for usage
+// with DagBuilderHelper with DagBuilderParams#NoCopy=true.
+func WithoutDatastore(bs blockstore.Blockstore) (blockstore.Blockstore, error) {
+	coercer := &dsCoercer{bs}
+
+	// the FileManager stores positional infos (positional mappings) in a
+	// datastore, which in our case is the blockstore coerced into a datastore.
+	fm := NewFileManager(coercer, "/")
```
Comment (Kubuxu) [inline blockstore.go:22]: So AFAIK the reason the FileManager takes a path is to prevent secret extraction.

Comment (raulk) [reply blockstore.go:22]: I have to confess that I naively thought this was a key prefix for the datastore entries. But it's a root path, and if this were a general utility, it should at least take the path as an arg. I'll close this PR, and we'll keep this component on the Filecoin side.

**Pull request closed without merging**